hadoop-hdds/docs/content/design/truncate.md

```diff
@@ -0,0 +1,146 @@
+ 
+The use cases for truncate include
+1. Transaction, journal handling for external systems. Transactions are appended to a   Ozone file while they succeed. When a transaction fails it is aborted and the file is truncated, which rolls it back to the previous successful transaction.
+2. Ability to truncate incomplete data added to the file as a result of a failed write. This particularly enables users to reuse the same files after a failed create or append.
+3. Allow users to remove data they accidentally appended.
+4. Improved support for FUSE and NFSv3 connectors to Ozone.
+
+In proposed approach `truncate` is performed only on a closed key. If the key is opened for write, an attempt to truncate fails.
+
+# Implementation
+
+## Timing diagram
+
+This timing diagram omit the delete block between SCM and Datanode.
+![avatar](image/truncate.png)
+
+## OM
+
+When OM receives `truncate(key, newLength)`,in the `keyTable`, OM deletes the blocks which are fully truncated, and updates the block length which is partially truncated.  
+
+OM stores `<key, RepeatedTruncateOmKeyInfo>` in truncateTable, and returns succ to client.  The `TruncateKey` in `RepeatedTruncateOmKeyInfo` represents one truncate key operation, so the list of `TruncateKey` allows us to store a list of truncate operations related to one key.
+
+OM starts a background thread to select a certain number of entries from `truncateTable`, and sends these entries to SCM by the `TruncateScmKeyRequestProto`.
+
+### Related New Class
+```
+// include a list of truncate operation
+public RepeatedTruncateOmKeyInfo {
+    private final String volumeName;
+    private final String bucketName;
+    // name of key client specified
+    private String keyName;
+    public List<TruncateKey> truncateKeyList;
+}
+```
+
+### Related New Proto
+```
+/**
+ * A truncate key request sent by OM to SCM, it contains
+ * multiple number of keys (and their blocks needed to be deleted and truncated).
+ */
+message TruncateScmKeyRequestProto {
+     repeated TruncateKey truncateKeys = 1;
+}
+
+/**
+ * A truncated key and its blocks needed to be deleted and the  block needed to be partially truncated.
+ */
+message TruncateKey {
+     required string key = 1;
+     repeated BlockID deleteBlocks = 2;
+     repeated PartialTruncateBlock partialTruncateBlocks = 3;
+}
+
+// A partially truncated block, including the newLength
+message PartialTruncateBlock {
+    required BlockID truncateBlock = 1;
+    required uint64 newLength = 2;
+}
+```
+
+## SCM
+When scm receives `TruncateScmKeyRequestProto`, for the `deleteBlocks`, scm has already implemented the code.
+
+For the `partialTruncateBlocks`, we process it as a transaction like `deleteBlocks`, store `<transactionId, TruncatedBlocksTransaction>` in `truncatedBlocksTable`, and return succ to OM.  
+
+We abstract the code related to delete block transaction, so that truncate and delete blocks can share the abstract code.
```

Comment:
SCM will hold on the delete blocks request for open container. Will the same logic be used on truncate blocks ?